### PR TITLE
Update our style guide

### PR DESCRIPTION
**Pull request by skullydazed**
## Description

Update our style guide to reflect a couple recent things that have been coming up.

* Officially move to 4 space indent
* Document our stance on `#ifdef` vs `#if defined()`

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by skullydazed** [comment]
Before we merge this I'd like to invite the community to weigh in. I know there are some 2 space fans out there but the majority of people seem to prefer 4 space indents, and a lot of our code is already indented that way.

**Comment by pelrun** [comment]
4 spaces? noooooooooo! :)

Ah, whatever. I'm happy enough just letting clang-format deal with it.

**Comment by drashna** [comment]
I'm going to need a bigger screen⸮

**Comment by alex-ong** [comment]
4 space master race wow it's me vs 3 others so far

**Comment by vomindoraan** [comment]
Could you summarize what the reasoning behind the switch from 2 to 4 is?

**Comment by nooges** [comment]
I used to be a 2 spacer, but do 4 now, as it’s visually easier to distinguish, especially since I like to use a relatively narrow font.

@skullydazed Can we also get some guidance on indentation of #ifdefs/#ifs? I’ve ran into different styles for that lately.

**Comment by Rionlion100** [comment]
4 spaces FTW for visual clarity and easy for beginners because this is also used by python

**Comment by pelrun** [comment]
Don't reference other things as an authority, as there's a strong example for every possible sensible combination you're after (also, Python happily uses whatever indentation you want, as long as it's consistent in the block.)

It's more important that whatever is chosen minimises the amount of work required. Right now, most of the codebase uses 4 spaces, and it would have to be entirely reformatted to change that.

The other concern is being able to automate most of the reformatting. It's far easier to maintain a style when nobody has to do it manually. But clang-format isn't perfectly configurable, so either you fight it constantly, or you get it as close as possible and then accept that as the baseline.

**Comment by vomindoraan** [comment]
A quick and dirty scan over the `.c` and `.h` files in the project produced this:

|Indentation  |# of files|
|-------------|----------|
|2 spaces     |1596      |
|3 spaces     |179       |
|4 spaces     |1070      |
|8 spaces     |44        |
|ambig. spaces|1203      |
|tab          |343       |
|none         |1010      |
|total        |5445      |

I used [this package](https://github.com/sindresorhus/detect-indent) for detecting indentation. Submodule files were ignored. The method I used isn't perfect, and there are more files that could've been ignored (e.g. the third-party files in `tmk_core/protocol/usb_hid`), but it gives a fairly accurate idea of the current state of indentation in the project.

With the objective part out of the way, I personally prefer 4 spaces for most projects. However, I also think that 2 spaces works well in QMK (keymap files are one example), and that a lot of files would need to be updated over time if the recommended indentation size were to change.

**Comment by XScorpion2** [comment]
2 spaces has made my eyes bleed. I welcome the 4 space overlords.
Now if we could just get with this century on bracket styling as well, that would be great.

File: docs/contributing.md [hunk 1]
```diff
@@ -71,6 +71,9 @@ Most of our style is pretty easy to pick up on, but right now it's not entirely
   * If you not sure if a comment is obvious, go ahead and include it.
 * In general we don't wrap lines, they can be as long as needed. If you do choose to wrap lines please do not wrap any wider than 76 columns.
 * We use `#pragma once` at the start of header files rather than old-style include guards (`#ifndef THIS_FILE_H`, `#define THIS_FILE_H`, ..., `#endif`)
+* We accept both forms of pre-processor if's: `#ifdef DEFINED` and `#if defined(DEFINED)`
```
Comment (vomindoraan) [inline docs/contributing.md]:
```suggestion
* We accept both forms of preprocessor if's: `#ifdef DEFINED` and `#if defined(DEFINED)`
```
I think “preprocessor” is the common form. I haven't seen it spelled “pre-processor” before.

**Comment by skullydazed** [comment]
Per @nooges' comment and after some discussion on discord I've updated the `#if` section a bit. There's still room to revise that if anyone has more comments.

File: docs/contributing.md [hunk 2]
```diff
@@ -71,6 +71,14 @@ Most of our style is pretty easy to pick up on, but right now it's not entirely
...
 * In general we don't wrap lines, they can be as long as needed. If you do choose to wrap lines please do not wrap any wider than 76 columns.
 * We use `#pragma once` at the start of header files rather than old-style include guards (`#ifndef THIS_FILE_H`, `#define THIS_FILE_H`, ..., `#endif`)
+* We accept both forms of preprocessor if's: `#ifdef DEFINED` and `#if defined(DEFINED)`
+  * If you are not sure which to prefer use the `#if defined(DEFINED)` form.
+  * Do not change existing code from one style to the other, except when moving to a multiple condition `#if`.
+  * Do not put whitespace between `#` and `if`.
+  * When deciding how (or if) to indent directives keep these points in mind:
+    * Readability is more important than consistency.
+    * Follow the file's existing style. If the file is mixed follow the style that makes sense for the section you are modifying.
+    * When choosing to indent you can follow the indention level of the surrounding C code, or preprocessor directives can have their own indent level. Choose the style that best communicates the intent of your code.
```
Comment (vomindoraan) [inline docs/contributing.md]: Here's the suggestion from Discord:
```suggestion
    * When choosing to indent you can follow the indention level of the surrounding C code, or preprocessor directives can have their own indent level. Choose the style that best communicates the intent of your code.
    * In either case, preprocessor indentation should not introduce additional levels of indentation to actual code. Otherwise, you can end up with code that's very hard to read (see [here](https://gist.github.com/vomindoraan/7934814ef7a6301f4ef1f462df3386df) for an example).
```
